Fix SnapshotBackwardsCompatibilityIT

nik9000 opened the pull request:
SnapshotBackwardsCompatibilityIT had drifted from its twin methods in
SharedClusterSnapshotRestoreIT. This creates SnapshotSharedTestCases which
makes the similar methods in both test cases use the same code, modulo the
cluster upgrade.

Related to #13522

nik9000 commented:
Rebased to resolve merge issues and renamed the shared test cases.

danielmitterdorfer commented on core/src/test/java/org/elasticsearch/snapshots/SnapshotSharedTest.java:
```diff
@@ -0,0 +1,239 @@
+            testCase.index("test-idx-3", "doc", Integer.toString(i), "foo", "baz" + i);
+        }
+        testCase.refresh();
+        assertHitCount(client().prepareCount("test-idx-1").get(), 100L);
```
`#prepareCount()`is deprecated. Do you think it makes sense to correct it already in this PR?

nik9000 replied:
Good call. I'll do that.

danielmitterdorfer commented on core/src/test/java/org/elasticsearch/snapshots/SnapshotSharedTest.java:
```diff
@@ -0,0 +1,239 @@
+/**
+ * Test cases shared between SharedClusterSnapshotRestoreIT and SnapshotBackwardsCompatibilityIT.
+ */
+public class SnapshotSharedTest {
```
Why did you choose delegation over inheritance in this case?

nik9000 replied:
Inheritance wouldn't work here because SharedClusterSnapshotRestoreIT and SnapshotBackwardsCompatibilityIT have different superclasses.

danielmitterdorfer replied:
Oh, I missed that. Thanks for clarifying.

danielmitterdorfer commented on core/src/test/java/org/elasticsearch/test/ESIntegTestCase.java:
```diff
@@ -1882,7 +1882,7 @@ public static DateTimeZone randomDateTimeZone() {
      * Returns path to a random directory that can be used to create a temporary file system repo
      */
-    public Path randomRepoPath() {
+    public static Path randomRepoPath() {
```
Why did you choose to make `#randomRepoPath()` static? As you pass in the test case instance to `SnapshotSharedTest`anyway? I assume you decided that because the implementation only relies on class state, not instance state.

nik9000 replied:
Yeah - at first I tried to go down the path of making everything static. I can just revert this though.

danielmitterdorfer commented on qa/backwards/shared/src/test/java/org/elasticsearch/test/ESBackcompatTestCase.java:
```diff
@@ -189,7 +189,7 @@ protected TestCluster buildTestCluster(Scope scope, long seed) throws IOExceptio
         logger.info("Build internal cluster at {}", new Object[] {cluster.httpAddresses()});
         ExternalNodeServiceClient nodeServiceClient = new ExternalNodeServiceClient();
-        ExternalNode externalNode = new ExternalNode(backwardsCompatibilityVersion(), randomLong(), nodeServiceClient, new NodeConfigurationSource() {
+        ExternalNode externalNode = new ExternalNode(backwardsCompatibilityVersion().toString(), randomLong(), nodeServiceClient, new NodeConfigurationSource() {
```
Would it make sense to change the signature of `ExternalNode` too, so it takes a `Version` parameter instead of a `String`? This would ensure at compile time that we pass really just a version instead of an arbitrary string.

nik9000 replied:
I kind of like keeping it a string because on the ExternalNodeService level "version" is just "suffix of the directory name into which elasticsearch was un-tar-ed." This seems like a reasonable spot to switch from Version to String.

danielmitterdorfer replied:
I see. I just tend to encode these constraints in the type system, that's why I've asked.

nik9000 replied:
Yeah, do that too. I guess what I mean is that this constraint doesn't actually exist to `ExternalNode`. Even naming the parameter version is a bit funky. Its really just a directory suffix.

danielmitterdorfer replied:
Ok, than that's fine for me. So overall LGTM.

danielmitterdorfer commented:
@nik9000: I left a few comments.

nik9000 commented:
Thanks @danielmitterdorfer! I've addressed the comments, squashed, and rebased. Since you gave it a LGTM in the comments I'll merge now.

Thanks against!

nik9000 commented:
s/against/again/!
